frozen_lake.py

```diff
@@ -0,0 +1,126 @@
+import numpy as np
+import sys
+from six import StringIO, b
+from gym import utils
+import discrete_env
+
+# Mapping between directions and index number
+LEFT = 0
+DOWN = 1
+RIGHT = 2
+UP = 3
+
+# Maps for the two different environments
+MAPS = {
+    "4x4": [
+        "SFFF",
+        "HFFG" "SFFFFFFF",
+        "FFFFFFFF",
+        "FFFHFFFF",
+        "FHFFFHFF",
+        "FFFHFFFF",
+        "FFHFFFHF",
+        "FHFFHFHF",
+        "FFFHFFFG",
+    ],
+}
+
+
+class FrozenLakeEnv(discrete_env.DiscreteEnv):
+    """
+    Winter is here. You and your friends were tossing around a frisbee at the park
+    when you made a wild throw that left the frisbee out in the middle of the lake.
+    The water is mostly frozen, but there are a few holes where the ice has melted.
+    If you step into one of those holes, you'll fall into the freezing water.
+    you navigate across the lake and retrieve the disc.
+    The surface is described using a grid like the following
+
+        SFFF
+        FHFH
+        FFFH
+        HFFG
+
+    S : starting point, safe
+    F : frozen surface, safe
+    H : hole, fall to your doom
+    G : goal, where the frisbee is located
+
+    The episode ends when you reach the goal or fall in a hole.
+    You receive a reward of 1 if you reach the goal, and zero otherwise.
+
+    """
+
+    metadata = {"render.modes": ["human", "ansi"]}
+
+    def __init__(self, desc=None, map_name="4x4", is_slippery=True):
+        if desc is None and map_name is None:
+            raise ValueError("Must provide either desc or map_name")
+        elif desc is None:
+            desc = MAPS[map_name]
+        self.desc = desc = np.asarray(desc, dtype="c")
+        self.nrow, self.ncol = nrow, ncol = desc.shape
+
+        nA = 4  # number of actions
+        nS = nrow * ncol  # number of states
+
+        isd = np.array(desc == b"S").astype("float64").ravel()
+        isd /= isd.sum()
+
+        P = {s: {a: [] for a in range(nA)} for s in range(nS)}
+
+        def to_s(row, col):
+            return row * ncol + col
+
+        def inc(row, col, a):
```

Comment:
Please follow the [docstring format in FastForest for functions](https://github.com/ThrunGroup/FastForest/blob/main/experiments/exp_utils.py#L12-L18)